temporal-sdk/src/test/java/io/temporal/worker/WorkerVersioningTest.java

```diff
@@ -0,0 +1,452 @@
+        TestWorkerVersioningMissingAnnotation.class);
+    defaultVersionWorker.start();
+
+    WorkerDeploymentVersion v1 =
+        new WorkerDeploymentVersion(testWorkflowRule.getDeploymentName(), "1.0");
+    DescribeWorkerDeploymentResponse describeResp1 = waitUntilWorkerDeploymentVisible(v1);
+    setCurrentVersion(v1, describeResp1.getConflictToken());
+
+    TestWorkflows.QueryableWorkflow wf1 =
+        testWorkflowRule.newWorkflowStubTimeoutOptions(
+            TestWorkflows.QueryableWorkflow.class, "default-versioning-behavior");
+    WorkflowExecution we1 = WorkflowClient.start(wf1::execute);
+    wf1.mySignal("done");
+    WorkflowClient workflowClient = testWorkflowRule.getWorkflowClient();
+    workflowClient.newUntypedWorkflowStub(we1.getWorkflowId()).getResult(String.class);
+
+    WorkflowExecutionHistory hist = testWorkflowRule.getExecutionHistory(we1.getWorkflowId());
+    Assert.assertTrue(
+        hist.getHistory().getEventsList().stream()
+            .anyMatch(
+                e ->
+                    e.getEventType() == EventType.EVENT_TYPE_WORKFLOW_TASK_COMPLETED
+                        && e.getWorkflowTaskCompletedEventAttributes().getVersioningBehavior()
+                            == io.temporal.api.enums.v1.VersioningBehavior
+                                .VERSIONING_BEHAVIOR_PINNED));
+  }
+
+  @WorkflowInterface
+  public interface DontAllowBehaviorAnnotationOnInterface {
+    @WorkflowMethod
+    @WorkflowVersioningBehavior(VersioningBehavior.VERSIONING_BEHAVIOR_PINNED)
+    void execute();
+  }
+
+  public static class DontAllowBehaviorAnnotationOnInterfaceImpl
+      implements DontAllowBehaviorAnnotationOnInterface {
+    @Override
+    public void execute() {}
+  }
+
+  @Test
+  public void testAnnotationNotAllowedOnInterface() {
+    IllegalArgumentException e =
+        Assert.assertThrows(
+            IllegalArgumentException.class,
+            () ->
+                testWorkflowRule
+                    .getWorker()
+                    .registerWorkflowImplementationTypes(
+                        DontAllowBehaviorAnnotationOnInterfaceImpl.class));
+    Assert.assertEquals(
+        "@WorkflowVersioningBehavior annotation is not allowed on interface methods, only on "
+            + "implementation methods: public abstract void io.temporal.worker.WorkerVersioningTest"
+            + "$DontAllowBehaviorAnnotationOnInterface.execute()",
+        e.getMessage());
+  }
+
+  private DescribeWorkerDeploymentResponse waitUntilWorkerDeploymentVisible(
+      WorkerDeploymentVersion v) {
+    DescribeWorkerDeploymentRequest req =
+        DescribeWorkerDeploymentRequest.newBuilder()
+            .setNamespace(testWorkflowRule.getTestEnvironment().getNamespace())
+            .setDeploymentName(v.getDeploymentName())
+            .build();
+    return Eventually.assertEventually(
+        Duration.ofSeconds(15),
+        () -> {
+          DescribeWorkerDeploymentResponse resp =
```

Comment:
Don't we need a high level deployment client? Go has that.

Reply:
No, we're not including that as part of the initial functionality